File: salmon_R/salmon_1.0_R/R/generatePGN.R
#' Protein-Gene Network Construction
#' 
#'  Generate protein-gene regulatory network from gene regulatory network and protein protein interaction network
#' @import slam
#'  
#' @param glist The vector of genes in the same order of the genes in log2FC data. The length of GList should be the same as the number of rows in log2FC data.
#' 
#' @param tftg The matrix of TF-gene interactions. The first column is the list of TFs, and the second column is the list of genes regulated by the corresponding TFs. The third column is optional, and if present, the column should contain the (confidence) score for each interaction.
#' 
#' @param ppi The matrix of protein-protein interactions. Each row of the first two columns give the protein pairs with interactions. The third column is optional, and if present, the column should contain the (confidence) score for each interaction.
#' 
#' @param tftg_thre A threshold for TF-gene interactions. This variable is used only when the confidence score of TF-gene interactions are given in the matrix tftg. Any TF-gene interactions with confidence scores lower than the threshold will be excluded.
#' 
#' @param ptf_thre A threshold for protein-TF interaction. This variable is used only when the confidence score of protein-TF interactions are given in the matrix ppi.  Any protein-TF interactions with the scores lower than the threshold will be excluded.
#' 
#' @param ppi_thre A threshold for protein-protein interaction. This variable is used only when the confidence score of protein-protein interactions are given in the matrix ppi.  Any protein-protein interactions with the scores lower than the threshold will be excluded.
#' 
#' @return \item{pgn}{The adjacency matrix of PGN}
#' 
#' @export
generatePGN <- function(glist,tftg,ppi,tftg_thre=NULL,ptf_thre=NULL,ppi_thre=NULL){
  glist <- sapply(glist, toupper)
  
  if (ncol(tftg)==3){
  names(tftg) <- c("tf","tg","score")
  }else names(tftg) <- c("tf","tg")
  if (ncol(ppi)==3){
    names(ppi) <- c("protein1","protein2","score")
  }else names(ppi) <- c("protein1","protein2")
  
  
  #GRN construction
  if(ncol(tftg)==3&&!is.null(tftg_thre)){
    tftg <- tftg[which(tftg$score>=tftg_thre),]
  }
  tftg <- as.data.frame(sapply(tftg, toupper))
  
  edge <- matrix(nrow = nrow(tftg),ncol = 3)
  edge[,1] <- match(tftg$tf,glist)
  edge[,2] <- match(tftg$tg,glist)
  if(ncol(tftg)==3){
    edge[,3] <- as.numeric(as.character(tftg$score))
  }else edge[,3] <- 1
  edge <- edge[!is.na(edge[,1])&!is.na(edge[,2]),]
  
  grn <- matrix(0,nrow = length(glist),ncol = length(glist))
  grn[edge[,1:2]] <- edge[,3]
  grn <- as.simple_triplet_matrix(grn)
  
  tfn <- length(unique(edge[,1]))
  dgn <- length(unique(edge[,2]))
  edgen <- length(grn$i)
  cat(sprintf("TF-Gene network with %d TFs, %d genes and %d interactions has been generated.\n",tfn,dgn,edgen))
  
  #PPI construction
  ppi <- as.data.frame(sapply(ppi, toupper))
  
  edge <- matrix(nrow = nrow(ppi),ncol = 3)
  edge[,1] <- match(ppi$protein1,glist)
  edge[,2] <- match(ppi$protein2,glist)
  if(ncol(ppi)==3){
    edge[,3] <- as.numeric(as.character(ppi$score))
  }else edge[,3] <- 1
  edge <- edge[which(edge[,1]!=edge[,2]&!is.na(edge[,1])&!is.na(edge[,2])),]
  
  first2columns <- as.data.frame(t(apply(edge[,1:2], 1, sort)))
  first2columns$V3 <- edge[,3]
  first2columns <- first2columns[order(-first2columns$V1,-first2columns$V2,-first2columns$V3),]
  dup <- duplicated(first2columns[,1:2])
  ppi_idx <- first2columns[!dup,]
  colnames(ppi_idx) <- c("V1","V2","V3")
  
  #Protein1-TF network
  tf_idx <- unique(grn$i)
  e1 <- ppi_idx[,1]%in%tf_idx|ppi_idx[,2]%in%tf_idx
  ppi_ly1 <- ppi_idx[e1,]
  proti1 <- setdiff(unique(c(ppi_ly1[,1],ppi_ly1[,2])),tf_idx)
  
  #set the order (prot1-TF)
  x <- ppi_ly1[,1]%in%tf_idx
  ppi_ly1[x,1:2] <- ppi_ly1[x,2:1]
  
  #copy TF2-TF1 interaction
  copyi <- ppi_ly1[,1]%in%tf_idx&ppi_ly1[,2]%in%tf_idx
  copy <- ppi_ly1[copyi,c(2,1,3)]
  colnames(copy) <- c("V1","V2","V3")
  ppi_ly1 <- rbind(ppi_ly1,copy)
  if(!is.null(ptf_thre)){
    ppi_ly1 <- ppi_ly1[which(ppi_ly1[,3]>=ptf_thre),]
  }
  
  ppn1 <- matrix(0,nrow = length(glist),ncol = length(glist))
  ppn1[as.matrix(ppi_ly1[,1:2])] <- ppi_ly1[,3]
  ppn1 <- as.simple_triplet_matrix(ppn1)
  
  #summary
  prot1n <- length(unique(ppi_ly1[,1]))
  tfn <- length(unique(ppi_ly1[,2]))
  edgen <- length(ppn1$i)
  cat(sprintf("Protein-TF network with %d proteins, %d TFs and %d interactions has been generated.\n",prot1n,tfn,edgen))
  
  #Prot2-Prot1 network
  e2 <- ppi_idx[,1]%in%proti1|ppi_idx[,2]%in%proti1
  e2 <- setdiff(which(e2),which(e1))
  ppi_ly2 <- ppi_idx[e2,]
  proti2 <- setdiff(unique(c(ppi_ly2[,1],ppi_ly2[,2])),proti1)
  
  #set the order(Prot2-Prot1)
  y <- ppi_ly2[,1]%in%proti1
  ppi_ly2[y,1:2] <- ppi_ly2[y,2:1]
  
  #copy Prot1b-Prot1a interaction
  copyi <- ppi_ly2[,1]%in%proti1&ppi_ly2[,2]%in%proti1
  copy <- ppi_ly2[copyi,c(2,1,3)]
  colnames(copy) <- c("V1","V2","V3")
  ppi_ly2 <- rbind(ppi_ly2,copy)
  if(!is.null(ppi_thre)){
    ppi_ly2 <- ppi_ly2[which(ppi_ly2$V3>=ppi_thre),]
  }
  
  ppn2 <- matrix(0,nrow = length(glist),ncol = length(glist))
  ppn2[as.matrix(ppi_ly2[,1:2])] <- ppi_ly2[,3]
  ppn2 <- as.simple_triplet_matrix(ppn2)
  
  #summary
  prot2n <- length(unique(ppi_ly2[,1]))
  prot1n <- length(unique(ppi_ly2[,2]))
  edgen <- length(ppn2$i)
  cat(sprintf("Protein-protein network with %d upstream proteins, %d proteins and %d interactions has been generated.\n",prot2n,prot1n,edgen))
  
  #Protein-gene network construction
  pgn <- crossprod_simple_triplet_matrix(as.simple_triplet_matrix(t(ppn2+diag(length(glist)))),ppn1)
  pgn <- crossprod_simple_triplet_matrix(as.simple_triplet_matrix(t(pgn+diag(length(glist)))),grn)
  pgn <- as.simple_triplet_matrix(pgn)
  
  #summary
  protn <- length(unique(pgn$i))
  dgn <- length(unique(pgn$j))
  edgen <- length(pgn$v)
  cat(sprintf("Protein-gene network with %d proteins and TFs, %d downstream genes and %d interactions has been generated.\n",protn,dgn,edgen))
  
  return(pgn)
}

File: salmon_MATLAB/salmon_1.0_MAT/generatePGN.m
function [pgn,ppn2] = generatePGN(GList,tftg,ppi,ppi_thre,tftg_thre,ptf_thre)

%-------------------------------------------------------------------------
%DESCRIPTION:
%           Generate protein-gene regulatory network from gene regulatory 
%           network and protein protein interaction network
%
%INPUT ARGUMENTS:
%glist           gene ID file. The tfnet and glist file should use the same 
%                system for gene names. The genes in glist should have the 
%                same order as the expression data.
%pdi             gene regulatory network. The first column is the gene names 
%                of transcription factors and the second column is gene being 
%                regulated; the third column is optional, if present, it is 
%                the score for each regulation.
%ppi             protein protein interaction network. The first two columns 
%                are the interacting proteins; the third column is optional, 
%                if present, it is the score for each interaction.
%ppi_thre        threshold for protein-protein interaction. This only has 
%                effect if ppi has a third score column. Interactions with 
%                score lower than the threshold will be ignored.
%tftg_thre       threshold for transcription factor-gene regulation. This 
%                only has effect if pdi has a third score column. Interactions 
%                with score lower than the threshold will be ignored.
%ptf_thre        threshold for protein-transcription factor interaction. This 
%                only has effect if ppi has a third score column. Interactions 
%                with score lower than the threshold will be ignored. 
%
%OUTPUT ARGUMENTS:
%pgn             a sparse matrix representing protein gene regulatory network.
%--------------------------------------------------------------------------

    if nargin<4 || isempty(ppi_thre)
        ppi_thre = 0;
    end
    
    if nargin<5 || isempty(tftg_thre)
        tftg_thre = 0;
    end
    
    if nargin<6 || isempty(ptf_thre)
        ptf_thre = 0;
    end
    
    n = length(GList);
    
    %%%%%%%%%%%%%%%%%%%%%%%% GRN CONSTRUCTION %%%%%%%%%%%%%%%%%%%%%%%%
    %%% Check whether tftg has 3rd column (= score)
    if size(tftg,2)<3
        tftg = [tftg,num2cell(ones(size(tftg,1),1))];
        tftg_thre = 0;
    end
    
    %%%% Trim TF-gene interactions (high score & existing in GList)
    tftg = tftg(cell2mat(tftg(:,3)) >= tftg_thre,:);
    mem1 = ismember(upper(tftg(:,1)),upper(GList));
    mem2 = ismember(upper(tftg(:,2)),upper(GList));
    tftg = tftg(mem1 & mem2,:);

    %%%% GRN construction
    grn = zeros(n);
    tf = unique(tftg(:,1));

    for j=1:length(tf)
        [~,tgi]=ismember(upper(tftg(strcmpi(tf(j),tftg(:,1)),2)),upper(GList));
        grn(strcmpi(tf(j),GList),tgi) = 1;
    end

    grn = sparse(grn);

    %%%% summary %%%%
    tfn = length(tf);
    dgn = nnz(sum(grn~=0,1));
    edgen = nnz(grn);
    fprintf('TF-Gene network (%d TFs, %d genes, %d interactions) has been generated.\n',tfn,dgn,edgen);
    
    %%%%%%%%%%%%%%%%%%% PPI NETWORK CONSTRUCTION %%%%%%%%%%%%%%%%%%%%
    %%% Check whether tftg has 3rd column (= score)
    if size(ppi,2)<3
       ppi = [ppi,num2cell(ones(size(ppi,1),1))];
       ppi_thre = 0;
    end

    %%%% Trim PPIs (high score & existing in GList)
    mem1 = ismember(upper(ppi(:,1)),upper(GList));
    mem2 = ismember(upper(ppi(:,2)),upper(GList));
    ppi = ppi(mem1 & mem2,:);

    % index of ppi
    [~,ic1] = ismember(upper(ppi(:,1)),upper(GList));
    [~,ic2] = ismember(upper(ppi(:,2)),upper(GList));
    ppi_idx = [ic1,ic2,cell2mat(ppi(:,3:end))];

    % Remove duplicated interactions with lower scores
    ppi_idx(:,1:2) = sort(ppi_idx(:,1:2),2);
    ppi_idx = flipud(sortrows(ppi_idx));
    [~,uni] = unique(ppi_idx(:,1:2),'rows','stable');
    ppi_idx = flipud(ppi_idx(uni,:));

    % Remove self-loops
    selfloop = ppi_idx(:,1)-ppi_idx(:,2)==0;
    ppi_idx(selfloop,:) = [];


    %%%% Prot1 - TF network %%%%
    tf_idx = find(sum(grn~=0,2));
    e1 = find(ismember(ppi_idx(:,1),tf_idx) | ismember(ppi_idx(:,2),tf_idx));
    ppi_ly1 = ppi_idx(e1,:);
    proti1 = setdiff(unique(ppi_ly1(:,1:2)),tf_idx);

    % set the order (prot1-TF)
    x = find(ismember(ppi_ly1(:,1),tf_idx));
    ppi_ly1(x,[1,2]) = ppi_ly1(x,[2,1]);

    % If there is a TF1-TF2 interaction, copy as TF2-TF1 as well.
    copyi = ismember(ppi_ly1(:,1),tf_idx) & ismember(ppi_ly1(:,2),tf_idx);
    ppi_ly1 = [ppi_ly1; ppi_ly1(copyi,[2,1,3:end])];

    ppn1 = zeros(n);
    tfi2 = unique(ppi_ly1(:,2));
    for j=1:length(tfi2)
        ind = find(ppi_ly1(:,2)==tfi2(j));
        ppn1(ppi_ly1(ind,1),tfi2(j)) = ppi_ly1(ind,3);
    end
    
    ppn1(ppn1<ptf_thre)=0;
    ppn1 = sparse(ppn1);

    %%%% summary %%%%
    prot1n = nnz(sum(ppn1~=0,2));
    tfn = length(tfi2);
    edgen = nnz(ppn1);
    fprintf('Protein-TF network (%d proteins, %d TFs, %d interactions) has been generated.\n',prot1n,tfn,edgen);
    

    %%%% Prot2 - Prot1 network %%%%
    e2 = find(ismember(ppi_idx(:,1),proti1) | ismember(ppi_idx(:,2),proti1));
    e2 = setdiff(e2,e1);
    ppi_ly2 = ppi_idx(e2,:);

    % set the order (prot2-prot1)
    y = find(ismember(ppi_ly2(:,1),proti1));
    ppi_ly2(y,[1,2]) = ppi_ly2(y,[2,1]);

    % If there is Prot1a-prot1b interactions, copy as Prot1b-Prot1a as well.
    copyi = ismember(ppi_ly2(:,1),proti1) & ismember(ppi_ly2(:,2),proti1);
    ppi_ly2 = [ppi_ly2; ppi_ly2(copyi,[2,1,3])];

    ppn2 = zeros(n);
    proti1b = unique(ppi_ly2(:,2));
    for j=1:length(proti1b)
        ind = find(ppi_ly2(:,2)==proti1b(j));
        ppn2(ppi_ly2(ind,1),proti1b(j)) = ppi_ly2(ind,3);
    end
    
    ppn2(ppn2<ppi_thre)=0;
    ppn2 = sparse(ppn2);

    %%%% summary %%%%
    prot2n = nnz(sum(ppn2~=0,2));
    prot1n = nnz(sum(ppn2~=0,1));
    edgen = nnz(ppn2);
    fprintf('Protein-Protein network (%d upstream proteins, %d proteins, %d interactions) has been generated.\n',prot2n,prot1n,edgen);

    %%
    %%%%%%%%%%%%%%%%%%% PROTEIN-GENE NETWORK CONSTRUCTION %%%%%%%%%%%%%%%%%%%%
    pgn = (ppn2+eye(n))*ppn1;
    pgn = (pgn + eye(n))*grn;

    %%%% summary %%%%
    protn = nnz(sum(pgn~=0,2));
    dgn = nnz(sum(pgn~=0,1));
    edgen = nnz(pgn);
    fprintf('Protein-Gene network (%d proteins and TFs, %d downstream genes, %d interactions) has been generated.\n',protn,dgn,edgen);

    pgn = sparse(pgn);    
end